src/potentials.h

```diff
@@ -680,53 +680,77 @@ class FunctorPotential : public PairPotentialBase {
+ * the EXACT energy, while if above ZERO is returned.
  */
-class TabulatedPotential : public FunctorPotential {
-
-    // expand spline data class to hold information about
-    // the sign of values for r<rmin
-    struct KnotData : public Tabulate::TabulatorBase<double>::data {
-        typedef Tabulate::TabulatorBase<double>::data base;
-        bool isNegativeBelowRmin = false;
+class SplinedPotential : public FunctorPotential {
+    /** @brief Expand spline data class to hold information about the sign of values for r<rmin */
+    class KnotData : public Tabulate::TabulatorBase<double>::data {
+      public:
+        using base = Tabulate::TabulatorBase<double>::data;
+        bool hardsphere_repulsion = false; //!< Use hardsphere repulsion for r smaller than rmin
         KnotData() = default;
-        inline KnotData(const base &b) : base(b) {}
+        KnotData(const base &);
     };
-    PairMatrix<KnotData> matrix_of_knots; // matrix with tabulated potential for each atom pair; cannot be Eigen Matrix
-    Tabulate::Andrea<double> spline;      // spline class
-    bool hardsphere = false;          // use hardsphere for r<rmin?
+
+    PairMatrix<KnotData> matrix_of_knots;               //!< Matrix with tabulated potential for each atom pair
+    Tabulate::Andrea<double> spline;                    //!< Spline method
+    bool hardsphere_repulsion = false;                  //!< Use hardsphere repulsion for r smaller than rmin
+    const int max_iterations = 1e6;                     //!< Max number of iterations when determining spline interval
+    void save_potentials();                             //!< Save splined and exact pair potentials to disk
+    double findLowerDistance(int, int, double, double); //!< Find lower distance for splining (rmin)
+    double findUpperDistance(int, int, double, double); //!< Find upper distance for splining (rmax)
+    double dr = 1e-2;                                   //!< Distance interval when searching for rmin and rmax
+    void createKnots(int, int, double, double);         //!< Create spline knots for pair of particles in [rmin:rmax]
 
   public:
-    TabulatedPotential(const std::string &name = "splined") : FunctorPotential(name) {};
+    explicit SplinedPotential(const std::string &name = "splined");
 
     /**
-     * @todo Simply return exact energy if `r2<=knots.rmin2`?
+     * Policies:
+     *
+     * 1. return splined potential if rmin>r<rmax
+     * 2. return zero if r>=rmax
+     * 3. return infinity if r<=rmin AND `hardsphere_repulsion` has been set to true
+     * 4. return exact energy if r<=rmin
      */
-    inline double operator()(const Particle &a, const Particle &b, double r2, const Point &) const override {
-        const KnotData &knots = matrix_of_knots(a.id, b.id);
-        if (r2 >= knots.rmax2)
+    inline double operator()(const Particle &p1, const Particle &p2, double r2, const Point &) const override {
```

Comment:
I prefer `if`-`else if`-`else` code flow for readability. I generally use only a single return statement with a return variable in the function unless there is a performance hit.

Reply:
clang-tidy discourages`if-else` after a call to return. In any case